src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,317 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        internal static int _disposedTensorCount;
+        internal static int _disposedCount;
+
+        static DisposeScopeManager()
+        {
+            torch.Tensor.OnTensorCreateCallback.Add(TryRegisterInDisposeScope);
+            torch.Tensor.OnTensorDisposeCallback.Add(TryUnregisterInDisposeScope);
+        }
+
+        [ThreadStatic] private static DisposeScopeManager _singleton;
+
+        public static DisposeScopeManager Singleton = (_singleton ??= new DisposeScopeManager());
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public static int DisposedTensorCount => _disposedTensorCount;
+        public static int DisposedCount => _disposedCount;
+
+        public static long TotalAllocatedSize => Singleton.InnerTotalAllocatedSize;
+
+        internal static void TryUnregisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryUnregisterInDisposeScope(disposable);
+
+        internal static void TryRegisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryRegisterInDisposeScope(disposable);
+
+        private long InnerTotalAllocatedSize {
+            get {
+                long size = 0;
+                foreach (var disposeScope in DisposeScopeStack) {
+                    size += disposeScope.TotalAllocatedSize;
+                }
+
+                return size;
+            }
+        }
+
+        private void InnerTryRegisterInDisposeScope(IDisposable disposable)
+        {
+            if (DisposeScopeStack.Count > 0) {
+                DisposeScopeStack.Peek().Include(disposable);
+            }
+        }
+
+        private void InnerTryUnregisterInDisposeScope(IDisposable disposable)
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
+                return;
+            }
+
+            foreach (var disposeScope in DisposeScopeStack) {
+                bool wasRemoved = disposeScope.Disposables.Remove(disposable);
+                if (wasRemoved && !disposeScope.DontReportOnExternalDisposes) {
+                    // File.AppendAllText(@"c:\slask\disp.txt",
```

Comment:
Commented code -- usually a no-no in my book. With it commented, the whole if-statement is redundant. Also, see my comment on 'DontReportOnExternalDisposes'

Reply:
Yep, it's temporary for the dispose investigation. But I'll make sure to remove it.

Reply:
Fixed